File: telegram_terminal_bot.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os/exec"
	"runtime"
	"strings"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"github.com/shirou/gopsutil/v3/cpu"
	"github.com/shirou/gopsutil/v3/disk"
	"github.com/shirou/gopsutil/v3/host"
	"github.com/shirou/gopsutil/v3/mem"
	"github.com/shirou/gopsutil/v3/net"
	"github.com/shirou/gopsutil/v3/process"
)

// Configuração
type Config struct {
	Token              string   `json:"token"`
	DonoUsername       string   `json:"dono_username"`
	IdsAutorizados     []int64  `json:"ids_autorizados"`
	UsuariosBloqueados []int64  `json:"usuarios_bloqueados"`
	TentativasMaximas  int      `json:"tentativas_maximas"`
}

var (
	config         Config
	tentativasFalhas = make(map[int64]int)
)

func main() {
	// Carregar configuração
	configFile, err := ioutil.ReadFile("config.json")
	if err != nil {
		log.Fatal("Erro ao ler config.json:", err)
	}

	err = json.Unmarshal(configFile, &config)
	if err != nil {
		log.Fatal("Erro ao parsear config.json:", err)
	}

	// Inicializar bot
	bot, err := tgbotapi.NewBotAPI(config.Token)
	if err != nil {
		log.Fatal("Erro ao inicializar bot:", err)
	}

	log.Printf("🤖 BOT-T-Terminal (Go) iniciado! @%s", bot.Self.UserName)

	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates := bot.GetUpdatesChan(u)

	for update := range updates {
		if update.Message == nil {
			continue
		}

		msg := update.Message

		// Verificar autorização
		if !verificarAutorizacao(bot, msg) {
			continue
		}

		// Processar comandos
		switch msg.Command() {
		case "start":
			enviarMensagemStart(bot, msg)
		case "cmd":
			executarComando(bot, msg)
		case "status":
			enviarStatus(bot, msg)
		case "processos":
			listarProcessos(bot, msg)
		case "memoria":
			enviarInfoMemoria(bot, msg)
		case "disco":
			enviarInfoDisco(bot, msg)
		case "rede":
			enviarInfoRede(bot, msg)
		case "ajuda":
			enviarMensagemStart(bot, msg)
		}
	}
}

func verificarAutorizacao(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) bool {
	userId := msg.From.ID

	// Verificar se está bloqueado
	for _, id := range config.UsuariosBloqueados {
		if id == userId {
			reply := tgbotapi.NewMessage(msg.Chat.ID, "❌ Você está bloqueado! Contate o administrador.")
			bot.Send(reply)
			return false
		}
	}

	// Verificar se está autorizado
	autorizado := false
	for _, id := range config.IdsAutorizados {
		if id == userId {
			autorizado = true
			break
		}
	}

	if !autorizado {
		tentativasFalhas[userId]++
		if tentativasFalhas[userId] >= config.TentativasMaximas {
			config.UsuariosBloqueados = append(config.UsuariosBloqueados, userId)
			configJson, _ := json.MarshalIndent(config, "", "    ")
			ioutil.WriteFile("config.json", configJson, 0644)
			reply := tgbotapi.NewMessage(msg.Chat.ID, "🚫 Muitas tentativas! Você foi bloqueado.")
			bot.Send(reply)
		} else {
			reply := tgbotapi.NewMessage(msg.Chat.ID, "⚠️ Você não está autorizado!")
			bot.Send(reply)
		}
		return false
	}

	return true
}

func enviarMensagemStart(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	mensagem := fmt.Sprintf(`🤖 *Bem-vindo ao BOT-T-Terminal!*

Este bot permite controlar seu servidor via Telegram.

*Comandos disponíveis:*
/cmd - Executar comando no servidor
/status - Ver status do servidor
/processos - Listar processos
/memoria - Ver uso de memória
/disco - Ver uso do disco
/rede - Ver informações de rede
/ajuda - Mostrar esta mensagem

*Observações:*
- Apenas usuários autorizados podem usar o bot
- Após %d tentativas falhas, você será bloqueado
- Apenas o dono (@%s) pode desbloquear usuários

*Desenvolvido por:* JOAC
*Versão:* 1.0.0`, config.TentativasMaximas, config.DonoUsername)

	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
}

func executarComando(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	comando := strings.TrimPrefix(msg.Text, "/cmd ")
	if comando == "" {
		reply := tgbotapi.NewMessage(msg.Chat.ID, "⚠️ Uso: /cmd <comando>")
		bot.Send(reply)
		return
	}

	cmd := exec.Command("bash", "-c", comando)
	output, err := cmd.CombinedOutput()
	
	if err != nil {
		reply := tgbotapi.NewMessage(msg.Chat.ID, fmt.Sprintf("❌ Erro:\n%s", err))
		bot.Send(reply)
		return
	}

	mensagem := fmt.Sprintf("✅ Resultado:\n```\n%s\n```", string(output))
	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
}

func enviarStatus(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	hostInfo, _ := host.Info()
	cpuInfo, _ := cpu.Info()
	memInfo, _ := mem.VirtualMemory()

	mensagem := fmt.Sprintf(`📊 *Status do Servidor*

*Sistema:*
• OS: %s %s
• Arch: %s
• Uptime: %d horas

*CPU:*
• Modelo: %s
• Cores: %d
• Velocidade: %.2f GHz

*Memória:*
• Total: %.2f GB
• Livre: %.2f GB
• Uso: %.2f%%`,
		hostInfo.Platform, hostInfo.PlatformVersion,
		runtime.GOARCH,
		hostInfo.Uptime/3600,
		cpuInfo[0].ModelName,
		runtime.NumCPU(),
		cpuInfo[0].Mhz/1000,
		float64(memInfo.Total)/(1024*1024*1024),
		float64(memInfo.Free)/(1024*1024*1024),
		memInfo.UsedPercent)

	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
}

func listarProcessos(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	processes, err := process.Processes()
	if err != nil {
		reply := tgbotapi.NewMessage(msg.Chat.ID, "❌ Erro ao listar processos")
		bot.Send(reply)
		return
	}

	mensagem := "📋 *Top 10 Processos:*\n\n"
	
	type ProcessInfo struct {
		pid   int32
		name  string
		cpu   float64
		mem   float32
	}

	processInfos := make([]ProcessInfo, 0)
	
	for _, p := range processes {
		name, _ := p.Name()
		cpu, _ := p.CPUPercent()
		mem, _ := p.MemoryPercent()
		
		processInfos = append(processInfos, ProcessInfo{
			pid:  p.Pid,
			name: name,
			cpu:  cpu,
			mem:  mem,
		})
	}

	// Ordenar por CPU
	for i := 0; i < len(processInfos)-1; i++ {
		for j := i + 1; j < len(processInfos); j++ {
			if processInfos[j].cpu > processInfos[i].cpu {
				processInfos[i], processInfos[j] = processInfos[j], processInfos[i]
			}
		}
	}

	// Pegar top 10
	count := 0
	for _, p := range processInfos {
		if count >= 10 {
			break
		}
		mensagem += fmt.Sprintf("• %s (PID: %d)\n", p.name, p.pid)
		mensagem += fmt.Sprintf("  CPU: %.1f%% | RAM: %.1f MB\n\n", p.cpu, float64(p.mem)*1024/100)
		count++
	}

	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
}

func enviarInfoMemoria(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	v, _ := mem.VirtualMemory()
	s, _ := mem.SwapMemory()

	mensagem := fmt.Sprintf(`💾 *Uso de Memória*

*RAM:*
• Total: %.2f GB
• Usado: %.2f GB
• Livre: %.2f GB
• Uso: %.2f%%

*Swap:*
• Total: %.2f GB
• Usado: %.2f GB
• Livre: %.2f GB
• Uso: %.2f%%`,
		float64(v.Total)/(1024*1024*1024),
		float64(v.Used)/(1024*1024*1024),
		float64(v.Free)/(1024*1024*1024),
		v.UsedPercent,
		float64(s.Total)/(1024*1024*1024),
		float64(s.Used)/(1024*1024*1024),
		float64(s.Free)/(1024*1024*1024),
		s.UsedPercent)

	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
}

func enviarInfoDisco(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	partitions, err := disk.Partitions(false)
	if err != nil {
		reply := tgbotapi.NewMessage(msg.Chat.ID, "❌ Erro ao obter informações do disco")
		bot.Send(reply)
		return
	}

	mensagem := "💽 *Uso de Disco*\n\n"

	for _, partition := range partitions {
		usage, err := disk.Usage(partition.Mountpoint)
		if err != nil {
			continue
		}

		mensagem += fmt.Sprintf("*%s:*\n", partition.Mountpoint)
		mensagem += fmt.Sprintf("• Total: %.2f GB\n", float64(usage.Total)/(1024*1024*1024))
		mensagem += fmt.Sprintf("• Usado: %.2f GB\n", float64(usage.Used)/(1024*1024*1024))
		mensagem += fmt.Sprintf("• Livre: %.2f GB\n", float64(usage.Free)/(1024*1024*1024))
		mensagem += fmt.Sprintf("• Uso: %.2f%%\n\n", usage.UsedPercent)
	}

	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
}

func enviarInfoRede(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	interfaces, err := net.Interfaces()
	if err != nil {
		reply := tgbotapi.NewMessage(msg.Chat.ID, "❌ Erro ao obter informações de rede")
		bot.Send(reply)
		return
	}

	stats, _ := net.IOCounters(true)
	
	mensagem := "🌐 *Informações de Rede*\n\n"

	for _, iface := range interfaces {
		if len(iface.Addrs) > 0 {
			mensagem += fmt.Sprintf("*%s:*\n", iface.Name)
			for _, addr := range iface.Addrs {
				mensagem += fmt.Sprintf("• IP: %s\n", addr.Addr)
			}
			mensagem += fmt.Sprintf("• MAC: %s\n", iface.HardwareAddr)
			mensagem += fmt.Sprintf("• MTU: %d\n\n", iface.MTU)
		}
	}

	for _, stat := range stats {
		mensagem += fmt.Sprintf("*%s (Stats):*\n", stat.Name)
		mensagem += fmt.Sprintf("• RX: %.2f MB\n", float64(stat.BytesRecv)/(1024*1024))
		mensagem += fmt.Sprintf("• TX: %.2f MB\n\n", float64(stat.BytesSent)/(1024*1024))
	}

	reply := tgbotapi.NewMessage(msg.Chat.ID, mensagem)
	reply.ParseMode = "Markdown"
	bot.Send(reply)
} 

File: telegram_terminal_bot.js
const TelegramBot = require('node-telegram-bot-api');
const si = require('systeminformation');
const fs = require('fs');
const { exec } = require('child_process');
const { promisify } = require('util');

const execAsync = promisify(exec);

// Carregar configuração
let config;
try {
    config = JSON.parse(fs.readFileSync('config.json', 'utf8'));
} catch (err) {
    console.error('❌ Erro ao carregar config.json:', err);
    process.exit(1);
}

// Inicializar bot
const bot = new TelegramBot(config.token, { polling: true });

// Mapa para armazenar tentativas de acesso
const tentativasFalhas = new Map();

// Verificar autorização
function verificarAutorizacao(msg) {
    const userId = msg.from.id;

    // Verificar se está bloqueado
    if (config.usuarios_bloqueados.includes(userId)) {
        bot.sendMessage(msg.chat.id, '❌ Você está bloqueado! Contate o administrador.');
        return false;
    }

    // Verificar se está autorizado
    if (!config.ids_autorizados.includes(userId)) {
        tentativasFalhas.set(userId, (tentativasFalhas.get(userId) || 0) + 1);
        
        if (tentativasFalhas.get(userId) >= config.tentativas_maximas) {
            config.usuarios_bloqueados.push(userId);
            fs.writeFileSync('config.json', JSON.stringify(config, null, 4));
            bot.sendMessage(msg.chat.id, '🚫 Muitas tentativas! Você foi bloqueado.');
        } else {
            bot.sendMessage(msg.chat.id, '⚠️ Você não está autorizado!');
        }
        return false;
    }

    return true;
}

// Comando /start
bot.onText(/\/start/, async (msg) => {
    if (!verificarAutorizacao(msg)) return;

    const mensagem = `🤖 *Bem-vindo ao BOT-T-Terminal!*

Este bot permite controlar seu servidor via Telegram.

*Comandos disponíveis:*
/cmd - Executar comando no servidor
/status - Ver status do servidor
/processos - Listar processos
/memoria - Ver uso de memória
/disco - Ver uso do disco
/rede - Ver informações de rede
/ajuda - Mostrar esta mensagem

*Observações:*
- Apenas usuários autorizados podem usar o bot
- Após ${config.tentativas_maximas} tentativas falhas, você será bloqueado
- Apenas o dono (@${config.dono_username}) pode desbloquear usuários

*Desenvolvido por:* JOAC
*Versão:* 1.0.0`;

    bot.sendMessage(msg.chat.id, mensagem, { parse_mode: 'Markdown' });
});

// Comando /cmd
bot.onText(/\/cmd (.+)/, async (msg, match) => {
    if (!verificarAutorizacao(msg)) return;

    const comando = match[1];
    
    try {
        const { stdout, stderr } = await execAsync(comando);
        const output = stdout || stderr;
        bot.sendMessage(msg.chat.id, `✅ Resultado:\n\`\`\`\n${output}\n\`\`\``, { parse_mode: 'Markdown' });
    } catch (err) {
        bot.sendMessage(msg.chat.id, `❌ Erro:\n${err.message}`);
    }
});

// Comando /status
bot.onText(/\/status/, async (msg) => {
    if (!verificarAutorizacao(msg)) return;

    try {
        const [cpu, mem, os] = await Promise.all([
            si.cpu(),
            si.mem(),
            si.osInfo()
        ]);

        const uptime = Math.floor(os.uptime / 3600);
        const mensagem = `📊 *Status do Servidor*

*Sistema:*
• OS: ${os.distro} ${os.release}
• Arch: ${os.arch}
• Uptime: ${uptime} horas

*CPU:*
• Modelo: ${cpu.manufacturer} ${cpu.brand}
• Cores: ${cpu.cores}
• Velocidade: ${cpu.speed} GHz

*Memória:*
• Total: ${(mem.total / (1024 * 1024 * 1024)).toFixed(2)} GB
• Livre: ${(mem.free / (1024 * 1024 * 1024)).toFixed(2)} GB
• Uso: ${((mem.used / mem.total) * 100).toFixed(2)}%`;

        bot.sendMessage(msg.chat.id, mensagem, { parse_mode: 'Markdown' });
    } catch (err) {
        bot.sendMessage(msg.chat.id, `❌ Erro ao obter status: ${err.message}`);
    }
});

// Comando /processos
bot.onText(/\/processos/, async (msg) => {
    if (!verificarAutorizacao(msg)) return;

    try {
        const processos = await si.processes();
        
        let mensagem = '📋 *Top 10 Processos:*\n\n';
        
        // Ordenar por CPU
        const top10 = processos.list
            .sort((a, b) => b.cpu - a.cpu)
            .slice(0, 10);
            
        for (const proc of top10) {
            mensagem += `• ${proc.name} (PID: ${proc.pid})\n`;
            mensagem += `  CPU: ${proc.cpu.toFixed(1)}% | RAM: ${(proc.memRss / 1024 / 1024).toFixed(1)} MB\n\n`;
        }

        bot.sendMessage(msg.chat.id, mensagem, { parse_mode: 'Markdown' });
    } catch (err) {
        bot.sendMessage(msg.chat.id, `❌ Erro ao listar processos: ${err.message}`);
    }
});

// Comando /memoria
bot.onText(/\/memoria/, async (msg) => {
    if (!verificarAutorizacao(msg)) return;

    try {
        const [mem, swap] = await Promise.all([
            si.mem(),
            si.memLayout()
        ]);

        const mensagem = `💾 *Uso de Memória*

*RAM:*
• Total: ${(mem.total / (1024 * 1024 * 1024)).toFixed(2)} GB
• Usado: ${(mem.used / (1024 * 1024 * 1024)).toFixed(2)} GB
• Livre: ${(mem.free / (1024 * 1024 * 1024)).toFixed(2)} GB
• Uso: ${((mem.used / mem.total) * 100).toFixed(2)}%

*Swap:*
• Total: ${(mem.swaptotal / (1024 * 1024 * 1024)).toFixed(2)} GB
• Usado: ${(mem.swapused / (1024 * 1024 * 1024)).toFixed(2)} GB
• Livre: ${((mem.swaptotal - mem.swapused) / (1024 * 1024 * 1024)).toFixed(2)} GB
• Uso: ${((mem.swapused / mem.swaptotal) * 100).toFixed(2)}%`;

        bot.sendMessage(msg.chat.id, mensagem, { parse_mode: 'Markdown' });
    } catch (err) {
        bot.sendMessage(msg.chat.id, `❌ Erro ao obter informações de memória: ${err.message}`);
    }
});

// Comando /disco
bot.onText(/\/disco/, async (msg) => {
    if (!verificarAutorizacao(msg)) return;

    try {
        const discos = await si.fsSize();
        
        let mensagem = '💽 *Uso de Disco*\n\n';
        
        for (const disco of discos) {
            mensagem += `*${disco.mount}:*\n`;
            mensagem += `• Total: ${(disco.size / (1024 * 1024 * 1024)).toFixed(2)} GB\n`;
            mensagem += `• Usado: ${(disco.used / (1024 * 1024 * 1024)).toFixed(2)} GB\n`;
            mensagem += `• Livre: ${((disco.size - disco.used) / (1024 * 1024 * 1024)).toFixed(2)} GB\n`;
            mensagem += `• Uso: ${disco.use.toFixed(2)}%\n\n`;
        }

        bot.sendMessage(msg.chat.id, mensagem, { parse_mode: 'Markdown' });
    } catch (err) {
        bot.sendMessage(msg.chat.id, `❌ Erro ao obter informações de disco: ${err.message}`);
    }
});

// Comando /rede
bot.onText(/\/rede/, async (msg) => {
    if (!verificarAutorizacao(msg)) return;

    try {
        const [interfaces, stats] = await Promise.all([
            si.networkInterfaces(),
            si.networkStats()
        ]);
        
        let mensagem = '🌐 *Informações de Rede*\n\n';
        
        // Interfaces
        for (const iface of interfaces) {
            if (iface.ip4) {
                mensagem += `*${iface.iface}:*\n`;
                mensagem += `• IP: ${iface.ip4}\n`;
                if (iface.ip6) mensagem += `• IPv6: ${iface.ip6}\n`;
                mensagem += `• MAC: ${iface.mac}\n`;
                mensagem += `• MTU: ${iface.mtu}\n\n`;
            }
        }
        
        // Estatísticas
        for (const stat of stats) {
            mensagem += `*${stat.iface} (Stats):*\n`;
            mensagem += `• RX: ${(stat.rx_bytes / (1024 * 1024)).toFixed(2)} MB\n`;
            mensagem += `• TX: ${(stat.tx_bytes / (1024 * 1024)).toFixed(2)} MB\n\n`;
        }

        bot.sendMessage(msg.chat.id, mensagem, { parse_mode: 'Markdown' });
    } catch (err) {
        bot.sendMessage(msg.chat.id, `❌ Erro ao obter informações de rede: ${err.message}`);
    }
});

// Comando /ajuda
bot.onText(/\/ajuda/, (msg) => {
    if (!verificarAutorizacao(msg)) return;
    bot.emit('message', msg, ['/start']);
});

// Inicialização
console.log('🤖 BOT-T-Terminal (Node.js) iniciado!'); 